Portfolio/SystemDesign.md

```diff
@@ -131,30 +131,236 @@ The stats screen was implemented according to the UI wireframe, however we have
+
+* MQTT has become the standard for IoT communication, due to its flexibility and efficiency which made it an easy choice. Due to the fact that responses are received virtually instantaneously, it is the ideal choice to send data such as the user's live health bar (this is particularly useful in accurately showing when the user's Bean "dies"). 
+
+To implement the MQTT communication protocol in our IoT product, we chose to use [HiveMQ](https://www.hivemq.com). Our team established two topics, 'doordie_web' and 'doordie_steps'. Only one request type uses 'doordie_steps', which is detailed below.
+
+Due to variability of payload attributes and sizes (especially concerning challenges), we made the decision to make a unifying request "type" parameter to work around the MQTT broker maximum character limit. A list of all valid request types made between devices is found in [MQTT_request_types.txt](/Documentation/Mqtt_request_types.txt), this document was used by our group to ensure clear communication between subsystems. We will expand on the key communication protocols and the request types below. 
+
+### DESKTOP AND M5STACK 
+To send a user's profile data to the M5Stack from the database:
 ``` 
 {
-    "type": "push profile",
-    "user_name": "Mario",
-    "total_steps": 2200,
-    "remaining_sec": 2000
+    	"type": "push profile",
+    	"user_name": "Mario",
+    	"total_steps": 2200,
+    	"remaining_sec": 2000
 }
 ```
 This is only sent when M5 sends a request to the MQTT broker:
 ```
 {
-    "type": "pull profile",
+    	"type": "pull profile",
+    	"user_name": "Mario"
+}
+```
+
+Processing sends all challenges a user is enrolled in to be updated on their M5Stack:
+```
+{
+    "type": "push user challenges",
+    "user_name": "Mario",
+    "challenges": [
+    {
+        "challenge_id": "1",
+        "challenge_name": "10K Step Challenge",
+        "description": "stepstep",
+        "end_time": 1589500800,				
+        "step_goal": 10000,
+        "reward": 800
+    },
+  
+    {
+        "challenge_id": "2",
+        "challenge_name": "Challenge 2",
+        "description": "runrun",
+        "end_time": 1589500800,
+        "step_goal": 2000,
+        "reward": 200
+    }]
+}
+```
+
+This is sent when the M5Stack requests the challenges of the user: 
+```
+{
+   	"type": "M5 pull challenges",
+	"user_name": "Mario"
+}
+```
+
+To update the statistics of the user on the M5Stack: 
+
+```
+{
+	"type": "pull user stats",
+	"user_name": "Mario"
+}
+```
+
+The database uses the following request type to send stats to the M5Stack: 
+```
+{
+	"type": "push user stats",
+	"user_name": "Mario",
+	"daily_record": 10000,
+	"weekly_record": 39000,
+	"weekly_current": 18032
+}
+```
+
+The M5Stack uses this request type to increment one step in the database, but there will be no response from the database. It is only for this request type that the 'doordie_steps' topic is used: 
```

Comment:
Using a different topic for this because we expect steps to be send every second or even more frequently at times, so once the number of users starts to increase, this topic is going to get completely flooded with this type of message.